### PR TITLE
ES|QL: vectorize eval

### PR DESCRIPTION
**Pull request by luigidellaquila**
Optimize generation of evaluators, so that the JIT can more easily vectorize the execution of math and binary operations.

The optimization applies only when all the following conditions are met: 
- the inputs are vector blocks
- the function is single value (no `mv_*`)
- the operators don't have to handle exceptions

```
Benchmark                 (operation)   Before     After      Speedup
EvalBenchmark.run                 abs   1,416      0,513      63%
EvalBenchmark.run                 add   1,629      1,635      --
EvalBenchmark.run          add_double   1,114      0,310      72%
EvalBenchmark.run          date_trunc   1,537      0,901      41%
EvalBenchmark.run      equal_to_const   0,584      0,314      46%
EvalBenchmark.run  long_equal_to_long   0,746      0,606      18%
EvalBenchmark.run   long_equal_to_int   1,812      0,858      52%
EvalBenchmark.run              mv_min   2,234      2,238      --
EvalBenchmark.run    mv_min_ascending   1,370      1,401      --
EvalBenchmark.run               rlike  10,332      9,395       9%
```

Tested on an M1 MBP

### CODE REVIEW

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

**Comment by luigidellaquila** [comment]
@elasticmachine update branch

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java [hunk 1]
```diff
@@ -182,8 +183,38 @@ private MethodSpec realEval(boolean blockStyle) {
...
+            boolean vectorize = false;
+            String vectorBuilderName = null;
+            if (blockStyle == false && processFunction.warnExceptions.isEmpty() && processOutputsMultivalued == false) {
+                ClassName type = processFunction.resultDataType(false);
+                vectorize = type.simpleName().startsWith("BytesRef") == false;
+                if (vectorize) {
+                    String arrayType = type.simpleName()
+                        .substring(0, type.simpleName().length() - "Vector".length())
+                        .toLowerCase(Locale.ROOT);
+                    builder.addStatement(arrayType + "[] buffer = result.values()");
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:195]: I wonder if we could remove `nextIndex` from `Vector.FixedBuilder` and set the `appendFoo` methods take an offset.

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java [hunk 2]
```diff
@@ -182,8 +183,38 @@ private MethodSpec realEval(boolean blockStyle) {
...
+                }
+            }
             builder.beginControlFlow("position: for (int p = 0; p < positionCount; p++)");
-            {
+            if (vectorize) {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:199]: Will this one ever be true?

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java (hunk 1 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:195]: I was thinking [the same](https://github.com/elastic/elasticsearch/pull/109332#discussion_r1627981386), I'll give it a try

File: x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java (hunk 2 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/EvaluatorImplementer.java:199]: Line [190](https://github.com/elastic/elasticsearch/pull/109332/files/e03f318d2f1d8a379a86286d2d391dd8f7f6f47a#diff-f8d41f1685f1dc618e979b50575b98910207aa3794c18c0d5a7ec7c05202c684R190) set is to `true` for all the types apart from `BytesRef`. This is the core of the optimization

File: x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/GreatestBooleanEvaluator.java [hunk 3]
```diff
@@ -86,13 +86,15 @@ public BooleanBlock eval(int positionCount, BooleanBlock[] valuesBlocks) {
...
     try(BooleanVector.Builder result = driverContext.blockFactory().newBooleanVectorBuilder(positionCount)) {
       boolean[] valuesValues = new boolean[values.length];
+      boolean[] buffer = result.values();
       position: for (int p = 0; p < positionCount; p++) {
         // unpack valuesVectors into valuesValues
         for (int i = 0; i < valuesVectors.length; i++) {
           valuesValues[i] = valuesVectors[i].getBoolean(p);
         }
-        result.appendBoolean(Greatest.process(valuesValues));
+        buffer[p] = Greatest.process(valuesValues);
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/GreatestBooleanEvaluator.java:95]: :+1:

File: x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/GreatestLongEvaluator.java [hunk 4]
```diff
@@ -56,8 +56,8 @@ public Block eval(Page page) {
   }
 
   public LongBlock eval(int positionCount, LongBlock[] valuesBlocks) {
+    long[] valuesValues = new long[values.length];
```
Comment (luigidellaquila) [inline x-pack/plugin/esql/src/main/generated/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/GreatestLongEvaluator.java:59]: This needs to be fixed

**Comment by luigidellaquila** [comment]
Benchmarks using the new API for `*VectorFixedBulder`

```
Benchmark                 (operation)  Mode  Cnt   Score   Error  Units
EvalBenchmark.run                 abs  avgt    7   0,544 ± 0,002  ns/op
EvalBenchmark.run                 add  avgt    7   1,627 ± 0,011  ns/op
EvalBenchmark.run          add_double  avgt    7   0,294 ± 0,031  ns/op
EvalBenchmark.run          date_trunc  avgt    7   0,898 ± 0,002  ns/op
EvalBenchmark.run      equal_to_const  avgt    7   0,312 ± 0,001  ns/op
EvalBenchmark.run  long_equal_to_long  avgt    7   0,605 ± 0,006  ns/op
EvalBenchmark.run   long_equal_to_int  avgt    7   0,857 ± 0,002  ns/op
EvalBenchmark.run              mv_min  avgt    7   2,234 ± 0,004  ns/op
EvalBenchmark.run    mv_min_ascending  avgt    7   1,399 ± 0,005  ns/op
EvalBenchmark.run               rlike  avgt    7  10,146 ± 0,039  ns/op
```

The performance improvement is the same, but the solution is much more elegant.
It also allowed some further optimizations.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java [hunk 5]
```diff
@@ -25,6 +25,8 @@ final class LongVectorFixedBuilder implements LongVector.FixedBuilder {
      */
     private int nextIndex;
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:26]: Is it possible to drop this entirely? If, after your PR, it is still used, then we should do it in a followup.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java [hunk 6]
```diff
@@ -38,6 +40,12 @@ public LongVectorFixedBuilder appendLong(long value) {
         return this;
     }
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:41]: Is this one still used? I think we could change the interface and just never call this. It's always slower. If it's still used, let's wait for a followup.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java [hunk 7]
```diff
@@ -38,6 +40,12 @@ public LongVectorFixedBuilder appendLong(long value) {
...
 
+    @Override
+    public LongVectorFixedBuilder appendLong(long value, int idx) {
```
Comment (nik9000) [inline x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:44]: I hate to ask this, but could you reverse the argument order? That way it *looks* a little more like array assignment. It's a lot of changing things, but hopefully they are all generated.

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by nik9000** [comment]
One more round! Sorry about that. I just really want to swap the order of the arguments.

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java (hunk 6 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:41]: There are still a couple of places using it, but I can try to remove it.
My only concern it that it's the entry point of `Builder` interface, so I'd have to make it throw an `UnsupportedOperationException`, or change the hierarchy.

Let me try and see how it goes

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java (hunk 7 above)
Comment (luigidellaquila) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:44]: Yes, most of the code is generated, it should be easy ;-)

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java (hunk 6 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:41]: If there are users hold it until the next PR please! It's nice to have this one be "mostly generated code"

File: x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java (hunk 7 above)
Comment (nik9000) [reply x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/LongVectorFixedBuilder.java:44]: Thank you!

**Comment by alex-spies** [comment]
Heya, didn't review in depth but the general change looks good and makes sense to me.

I ran the eval bench on my i7-12700H (so we get some x86-64 coverage as well):
```
        (operation)  Mode  Cnt  Before  Error  After   Error  Units
                abs  avgt    7  1.404 ± 0.145  0.774 ± 0.007  ns/op
                add  avgt    7  1.905 ± 0.022  1.895 ± 0.023  ns/op
         add_double  avgt    7  -----   -----  0.684 ± 0.010  ns/op // No idea why this didn't run on main
         date_trunc  avgt    7  2.683 ± 0.012  2.670 ± 0.010  ns/op //  - maybe ran into an exception
     equal_to_const  avgt    7  0.585 ± 0.003  0.257 ± 0.003  ns/op
 long_equal_to_long  avgt    7  0.621 ± 0.008  0.362 ± 0.007  ns/op
  long_equal_to_int  avgt    7  1.945 ± 0.067  0.993 ± 0.007  ns/op
             mv_min  avgt    7  2.216 ± 0.015  2.198 ± 0.020  ns/op
   mv_min_ascending  avgt    7  1.390 ± 0.068  1.385 ± 0.084  ns/op
              rlike  avgt    7  8.590 ± 0.276  8.663 ± 0.634  ns/op
```
This looks consistent with @luigidellaquila 's results on M1 MBP.

**Comment by luigidellaquila** [comment]
Thanks a lot @alex-spies! 
So we have more or less the same improvements also on x86-64 arch, excellent!
Not for `date_trunc` though... strange...

```
add_double  avgt    7  -----   -----  0.684 ± 0.010  ns/op // No idea why this didn't run on main
```

Just because it doesn't exist on main 😉, I introduced it with this change to test math operations, because I knew I could not optimize the normal `add` (that uses ints and catches exceptions on overflows)

**Review by nik9000 — APPROVED** [review]

**Pull request merged**